File: backend/scripts/check_structure.py
"""Structure linter for backend/ — enforces the R1/R5/R7 folder-splitting rules.

Usage: python scripts/check_structure.py [--layer routers,services,...]
Exits non-zero if any violation is found.
"""
from __future__ import annotations

import argparse
import os
import re
import sys
from collections import defaultdict
from pathlib import Path

BACKEND_ROOT = Path(__file__).resolve().parent.parent

# Layers this check applies to (top-level packages under backend/).
LAYERS = [
    "routers", "services", "repositories", "application",
    "analytics", "domain", "scripts",
]

SKIP_DIR_NAMES = {"__pycache__", "data", "__init__.py"}

SIZE_THRESHOLD = 8       # R1(b): >8 source files (excl. __init__.py) in one folder
PREFIX_THRESHOLD = 3     # R1(a): >=3 files sharing a domain-ish prefix
TEST_FLAT_THRESHOLD = 25  # R1(c): >25 test files flat in one folder
MAX_DEPTH_UNDER_LAYER = 2  # R5: <layer>/<domain>/<subdomain>/file at most

# R7 — canonical backend domain names, and keywords that indicate a file
# belongs to that domain even when its folder name doesn't say so.
CANONICAL_DOMAINS = {
    "accounts", "content", "vocabulary_quiz", "learner_model",
    "research", "speech", "speaking", "classroom", "platform",
}
DOMAIN_KEYWORDS: dict[str, str] = {
    "bkt": "learner_model",
    "srs": "learner_model",
    "knowledge_analytics": "learner_model",
    "knowledge_tracing": "learner_model",
    "weak_words": "learner_model",
    "review_queue": "learner_model",
    "vocabulary_state": "learner_model",
    "vocab_quiz": "vocabulary_quiz",
    "quiz_review": "vocabulary_quiz",
    "quiz_attempt": "vocabulary_quiz",
    "research": "research",
    "vocabulary_research": "research",
    "placement_test": "research",
    "story": "content",
    "stories": "content",
    "vocabulary_import": "content",
    "vocabulary_audio_import": "content",
    "content_reset": "content",
    "content_verification": "content",
    "help_request": "classroom",
    "speaking_progress": "speaking",
    "submission": "speaking",
    "verified_speech": "speaking",
    "verified_speaking": "speaking",
    "audio_record": "speaking",
    "media": "platform",
    "health": "platform",
    "admin": "accounts",
    "student": "accounts",
    "teacher": "accounts",
}


def stem_prefix(stem: str) -> str:
    return stem.split("_")[0]


def find_prefix_clusters(stems: list[str]) -> dict[str, list[str]]:
    groups: dict[str, list[str]] = defaultdict(list)
    for stem in stems:
        groups[stem_prefix(stem)].append(stem)
    return {prefix: members for prefix, members in groups.items()
            if len(members) >= PREFIX_THRESHOLD}


def domain_keyword_hits(stem: str) -> set[str]:
    hits = set()
    for keyword, domain in DOMAIN_KEYWORDS.items():
        if keyword in stem:
            hits.add(domain)
    return hits


def iter_py_files(directory: Path):
    for entry in directory.iterdir():
        if entry.is_file() and entry.suffix == ".py":
            yield entry


def check_folder_size_and_prefix(directory: Path, violations: list[str]) -> None:
    source_files = [
        f for f in iter_py_files(directory)
        if f.name != "__init__.py" and not f.name.startswith("test_")
    ]
    if len(source_files) > SIZE_THRESHOLD:
        violations.append(
            f"R1(b) size: {rel(directory)} has {len(source_files)} source files "
            f"(threshold {SIZE_THRESHOLD}): "
            + ", ".join(sorted(f.name for f in source_files))
        )

    stems = [f.stem for f in source_files]
    for prefix, members in find_prefix_clusters(stems).items():
        violations.append(
            f"R1(a) prefix: {rel(directory)} has {len(members)} files sharing "
            f"prefix '{prefix}': {', '.join(sorted(members))}"
        )


def check_test_folder_flat(directory: Path, violations: list[str]) -> None:
    test_files = [f for f in iter_py_files(directory) if f.name.startswith("test_")]
    subdirs = [d for d in directory.iterdir() if d.is_dir() and d.name not in SKIP_DIR_NAMES]
    if len(test_files) > TEST_FLAT_THRESHOLD and not subdirs:
        violations.append(
            f"R1(c) flat tests: {rel(directory)} has {len(test_files)} test files "
            f"flat with no domain subfolders (threshold {TEST_FLAT_THRESHOLD})"
        )


def check_depth(layer_dir: Path, violations: list[str]) -> None:
    for dirpath, dirnames, _filenames in os.walk(layer_dir):
        dirnames[:] = [d for d in dirnames if d not in SKIP_DIR_NAMES and not d.startswith(".")]
        depth = len(Path(dirpath).relative_to(layer_dir).parts)
        if depth > MAX_DEPTH_UNDER_LAYER:
            violations.append(
                f"R5 depth: {rel(Path(dirpath))} is {depth} levels under "
                f"{layer_dir.name}/ (max {MAX_DEPTH_UNDER_LAYER})"
            )


def check_domain_naming(directory: Path, violations: list[str]) -> None:
    top_level_name = directory.name
    if top_level_name in CANONICAL_DOMAINS:
        return
    hits: dict[str, list[str]] = defaultdict(list)
    for f in iter_py_files(directory):
        if f.name == "__init__.py":
            continue
        for domain in domain_keyword_hits(f.stem):
            hits[domain].append(f.name)
    for domain, files in hits.items():
        violations.append(
            f"R7 naming: {rel(directory)} holds {len(files)} file(s) matching "
            f"domain '{domain}' but the folder isn't named for a canonical "
            f"domain: {', '.join(sorted(files))}"
        )


def rel(path: Path) -> str:
    return str(path.relative_to(BACKEND_ROOT)).replace("\\", "/")


def walk_layer(layer_dir: Path, violations: list[str]) -> None:
    if not layer_dir.is_dir():
        return
    check_depth(layer_dir, violations)
    for dirpath, dirnames, _filenames in os.walk(layer_dir):
        dirnames[:] = [d for d in dirnames if d not in SKIP_DIR_NAMES and not d.startswith(".")]
        current = Path(dirpath)
        if current.name == "tests" or "test" in current.name:
            check_test_folder_flat(current, violations)
            continue
        check_folder_size_and_prefix(current, violations)
        check_domain_naming(current, violations)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--layer", help="comma-separated subset of layers to check")
    args = parser.parse_args()

    layers = args.layer.split(",") if args.layer else LAYERS
    violations: list[str] = []

    for layer in layers:
        walk_layer(BACKEND_ROOT / layer, violations)

    tests_dir = BACKEND_ROOT / "tests"
    if tests_dir.is_dir():
        check_test_folder_flat(tests_dir, violations)

    if not violations:
        print("check_structure: no violations found.")
        return 0

    print(f"check_structure: {len(violations)} violation(s) found:\n")
    for v in sorted(violations):
        print(f"  - {v}")
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: frontend/scripts/check-structure.mjs
import { readdir, readFile } from "node:fs/promises";
import path from "node:path";

const src = path.resolve("src");
const LEGACY_SHIM_BASELINE = 0;
const allowedTopLevel = new Set([
  "app", "components", "config", "data", "entities", "features", "helpers",
  "hooks", "i18n", "integration", "services", "shared",
  "styles", "test", "types", "utils",
]);
const errors = [];
const warnings = [];
let shimCount = 0;

// R1/R5/R7 thresholds and domain table (see the pasted structure-rules spec).
const SIZE_THRESHOLD = 8;
const PREFIX_THRESHOLD = 3;
const TEST_FLAT_THRESHOLD = 25;
const MAX_DEPTH_UNDER_LAYER = 2;
const LAYER_ROOTS = new Set(["app", "features", "entities", "shared"]);
const CANONICAL_DOMAINS = new Set([
  "auth", "roster", "topic", "quiz", "mastery", "research", "speaking", "classroom", "shared",
]);
const DOMAIN_KEYWORDS = {
  quiz: "quiz",
  vocabulary: "topic",
  vocab: "quiz",
  story: "topic",
  stories: "topic",
  student: "roster",
  teacher: "roster",
  admin: "roster",
  roster: "roster",
  login: "auth",
  bkt: "mastery",
  srs: "mastery",
  mastery: "mastery",
  research: "research",
  placement: "research",
  speech: "speaking",
  conversation: "speaking",
  speaking: "speaking",
  submit: "speaking",
  help: "classroom",
  audio: "shared",
  media: "shared",
  health: "shared",
};

// Registry of every directory's own files, filled in during visit() and
// consumed once the walk is done (R1/R5/R7 need a full folder listing).
const dirRegistry = new Map();

function relative(file) {
  return path.relative(src, file).replaceAll(path.sep, "/");
}

function leadingWord(baseName) {
  const m = baseName.match(/^[A-Z]?[a-z0-9]+/);
  return (m ? m[0] : baseName).toLowerCase();
}

function findPrefixClusters(stems) {
  const groups = new Map();
  for (const stem of stems) {
    const key = leadingWord(stem);
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key).push(stem);
  }
  return [...groups.entries()].filter(([, members]) => members.length >= PREFIX_THRESHOLD);
}

function domainKeywordHits(stem) {
  const lower = stem.toLowerCase();
  const hits = new Set();
  for (const [keyword, domain] of Object.entries(DOMAIN_KEYWORDS)) {
    if (lower.includes(keyword)) hits.add(domain);
  }
  return hits;
}

async function visit(directory) {
  const entries = await readdir(directory, { withFileTypes: true });
  const rel = relative(directory);
  const record = { sourceFiles: [], testFiles: [], hasSubdirs: false };
  dirRegistry.set(rel, record);

  for (const entry of entries) {
    const fullPath = path.join(directory, entry.name);
    const entryRel = relative(fullPath);
    if (entry.isDirectory()) {
      record.hasSubdirs = true;
      if (directory === src && !allowedTopLevel.has(entry.name)) {
        errors.push(`${entryRel}: src/ contains an unapproved top-level directory`);
      }
      if (entry.name !== "src" && !/^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$/.test(entry.name)) {
        errors.push(`${entryRel}: directory names must be kebab-case`);
      }
      await visit(fullPath);
      continue;
    }
    if (!/\.(?:ts|tsx)$/.test(entry.name)) continue;

    const isTest = /\.test\.tsx?$/.test(entry.name);
    const isIndex = entry.name === "index.ts";
    if (isTest) record.testFiles.push(entry.name);
    else if (!isIndex && !entry.name.endsWith(".d.ts")) record.sourceFiles.push(entry.name);

    const text = await readFile(fullPath, "utf8");
    shimCount += (text.match(/LEGACY-SHIM/g) ?? []).length;
    if (/^(?:utils|helpers|misc|common)\.(?:tsx?|jsx?)$/.test(entry.name)) {
      errors.push(`${entryRel}: generic utility filenames are not allowed`);
    }
    if (/^[A-Z]\w*\.[A-Za-z]\w*\.tsx?$/.test(entry.name) && !entry.name.includes(".test.") && !entry.name.endsWith(".d.ts")) {
      errors.push(`${entryRel}: dotted pseudo-folders are not allowed`);
    }
    if (/\bfetch\s*\(/.test(text) && entry.name !== "api.ts" && !entryRel.startsWith("shared/api/")) {
      errors.push(`${entryRel}: fetch() belongs in an api.ts module or shared/api/`);
    }
    const lines = text.split(/\r?\n/).length;
    const limit = entry.name.endsWith(".tsx") ? 250 : 300;
    if (lines >= 500) errors.push(`${entryRel}: ${lines} lines (500-line limit)`);
    else if (lines >= limit) warnings.push(`${entryRel}: ${lines} lines (review at ${limit})`);
  }
}

function checkStructureRules() {
  for (const [rel, record] of dirRegistry) {
    if (rel === "") continue;
    const { sourceFiles, testFiles, hasSubdirs } = record;

    if (sourceFiles.length > SIZE_THRESHOLD) {
      errors.push(
        `R1(b) size: ${rel} has ${sourceFiles.length} source files (threshold ${SIZE_THRESHOLD}): `
        + sourceFiles.slice().sort().join(", ")
      );
    }
    for (const [prefix, members] of findPrefixClusters(sourceFiles.map((f) => f.replace(/\.tsx?$/, "")))) {
      errors.push(
        `R1(a) prefix: ${rel} has ${members.length} files sharing prefix '${prefix}': `
        + members.slice().sort().join(", ")
      );
    }
    if (testFiles.length > TEST_FLAT_THRESHOLD && !hasSubdirs) {
      errors.push(
        `R1(c) flat tests: ${rel} has ${testFiles.length} test files flat with no domain subfolders (threshold ${TEST_FLAT_THRESHOLD})`
      );
    }

    const topLevelName = rel.split("/")[0];
    if (!CANONICAL_DOMAINS.has(topLevelName)) {
      const hitsByDomain = new Map();
      for (const f of sourceFiles) {
        const stem = f.replace(/\.tsx?$/, "");
        for (const domain of domainKeywordHits(stem)) {
          if (!hitsByDomain.has(domain)) hitsByDomain.set(domain, []);
          hitsByDomain.get(domain).push(f);
        }
      }
      for (const [domain, files] of hitsByDomain) {
        errors.push(
          `R7 naming: ${rel} holds ${files.length} file(s) matching domain '${domain}' but the folder isn't named for a canonical domain: `
          + files.slice().sort().join(", ")
        );
      }
    }
  }

  for (const layerName of LAYER_ROOTS) {
    for (const rel of dirRegistry.keys()) {
      if (!rel.startsWith(`${layerName}/`)) continue;
      const depth = rel.split("/").length - 1; // path segments below the layer root
      if (depth > MAX_DEPTH_UNDER_LAYER) {
        errors.push(`R5 depth: ${rel} is ${depth} levels under ${layerName}/ (max ${MAX_DEPTH_UNDER_LAYER})`);
      }
    }
  }
}

await visit(src);
checkStructureRules();
if (shimCount > LEGACY_SHIM_BASELINE) errors.push(`LEGACY-SHIM count is ${shimCount}; baseline allows ${LEGACY_SHIM_BASELINE}`);
for (const warning of warnings) console.warn(`warning: ${warning}`);
for (const error of errors) console.error(`error: ${error}`);
if (errors.length) process.exit(1);
console.log(`structure ok (${shimCount} LEGACY-SHIM, ${warnings.length} review warning${warnings.length === 1 ? "" : "s"})`);
